### PR TITLE
Add Netty leak logger for throwing in promise completion callback

### PR DESCRIPTION
**Pull request by original-brownbear**
If we throw in the promise completion, even `Throwable` will just get logged by Netty. This uses the same mechanism we use to detect leaks to fail tests in that case as well. Noticed while working on using pooling in search responses and missing a bunch of double-free bugs that did not fail tests but showed in logs.

relates https://github.com/netty/netty/issues/6096

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 1]
```diff
@@ -263,13 +264,35 @@ public void append(LogEvent event) {
             };
             leakAppender.start();
             Loggers.addAppender(leakLogger, leakAppender);
-            // shutdown hook so that when the test JVM exits, logging is shutdown too
-            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
```
Comment (original-brownbear) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:267]: This shouldn't have been put in the loop, the last two lines of this hook need only run once.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 2]
```diff
@@ -263,13 +264,35 @@ public void append(LogEvent event) {
...
+        Logger promiseUncaughtLogger = LogManager.getLogger("io.netty.util.concurrent.DefaultPromise");
+        final Appender uncaughtAppender = new AbstractAppender(
+            promiseUncaughtLogger.getName(),
+            null,
+            PatternLayout.newBuilder().withPattern("%m").build()
+        ) {
+            @Override
+            public void append(LogEvent event) {
+                String message = event.getMessage().getFormattedMessage();
+                if (Level.WARN.equals(event.getLevel())) {
```
Comment (original-brownbear) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:278]: Any warning is an uncaught exception here, so no further checks needed.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 3]
```diff
@@ -263,13 +264,35 @@ public void append(LogEvent event) {
...
         }
+        Logger promiseUncaughtLogger = LogManager.getLogger("io.netty.util.concurrent.DefaultPromise");
+        final Appender uncaughtAppender = new AbstractAppender(
+            promiseUncaughtLogger.getName(),
+            null,
+            PatternLayout.newBuilder().withPattern("%m").build()
+        ) {
+            @Override
+            public void append(LogEvent event) {
+                String message = event.getMessage().getFormattedMessage();
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:277]: No need to format the message if it's not a warning?

**Review by DaveCTurner — APPROVED** [review]
LGTM (one nit) but also we should be failing the node on an `Error` in prod rather than just logging and continuing.

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java (hunk 3 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: I guess today this thing only logs warnings but we can definitely save a line here, move the formatting into the condition :)

**Pull request merged**